apps/voting/contracts/Voting.sol

```diff
@@ -422,6 +574,68 @@ contract Voting is IForwarder, AragonApp {
         emit ExecuteVote(_voteId);
     }
 
+    /**
+     * @dev internal function to add the `_voter` to the list of delegated voters for the `_delegate` and update the delegate for the `_voter`
+     * @param _delegate address of the delegate
+     * @param _voter address of the voter
+     */
+    function _addDelegatedAddressFor(address _delegate, address _voter) internal {
+        uint256 delegatedVotersCount = delegatedVoters[_delegate].addresses.length;
+        require(delegatedVotersCount <= UINT_96_MAX, ERROR_MAX_DELEGATED_VOTERS_REACHED);
+
+        delegatedVoters[_delegate].addresses.push(_voter);
+        delegates[_voter] = Delegate(_delegate, uint96(delegatedVotersCount));
+        emit SetDelegate(_voter, _delegate);
+    }
+
+    /**
+     * @dev internal function to remove  the `_voter` from the list of delegated voters for the `_delegate`
+     * @param _delegate address of the delegate
+     * @param _voter address of the voter
+     */
+    function _removeDelegatedAddressFor(address _delegate, address _voter) internal {
+        uint256 delegatedVotersCount = delegatedVoters[_delegate].addresses.length;
+        require(delegatedVotersCount > 0, ERROR_DELEGATE_NOT_SET);
+
+        uint96 voterIndex = delegates[_voter].voterIndex;
+        assert(delegatedVoters[_delegate].addresses[voterIndex] == _voter);
+        address lastVoter = delegatedVoters[_delegate].addresses[delegatedVotersCount - 1];
+        if (voterIndex < delegatedVotersCount - 1) {
+            delegatedVoters[_delegate].addresses[voterIndex] = lastVoter;
+            delegates[lastVoter].voterIndex = voterIndex;
+        }
+        delegatedVoters[_delegate].addresses.length--;
+        emit ResetDelegate(_voter, _delegate);
+    }
+
+    /**
+     * @notice Return the sliced list of delegated voters for `_delegate` and their voting power at the `_blockNumber`
+     * @param _delegate the address of the delegate
+     * @param _offset the number of delegated voters from the start of the list to skip
+     * @param _limit the number of delegated voters to return
+     * @return the array of delegated voters
+     * @return the array of voting power of delegated voters
+     */
+    function _getDelegatedVotersAt(address _delegate, uint256 _offset, uint256 _limit, uint256 _blockNumber) internal view returns (address[] memory votersList, uint256[] memory votingPowerList) {
```

Comment:
this approach may have a small negative impact if the addition/removal of a delegator occurs in between reading parts of the list.

e.g. there are 100 delegates in total, some delegate was at the beginning of the list, we read 50 records at a time, suppose a delegate calls resetDelegate between read requests. Since the last element of the array is placed in the place of the deleted one, the delegate from the end of the list will not be sampled at all.
Moreover, if the delegate manages to call resetDelegate and setDelegate repeatedly between requests, it will get into the selection 2 times

Reply:
I suggest to fetch from frontend the list of voters for the delegate and than call a newly introduced method `balanceOfAtMultiple` - so effectively split this into two view methods. This would simplify the code and reduce the logic in the contract.

Reply:
I split this getter into two separate functions, `_getVotingPowerMultipleAt` and `getDelegatedVoters`. Addressed in cd8697b8dbccbd21a4caa99e619bcf7889dd9e28.